File: unity-sdk/Packages/com.bugreporter.sdk/Runtime/ClipRecorder.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using UnityEngine;
using UnityEngine.Experimental.Rendering;
using UnityEngine.Rendering;

namespace BugReporter
{
    /// <summary>
    /// Keeps a rolling buffer of the last N seconds of the screen as small JPEG frames (a "flipbook"), so a
    /// report can carry a short clip of what led to the bug. Capture is throttled to a low fps and uses an
    /// async GPU readback so it doesn't stall the render thread. Opt-in (<see cref="BugReporterConfig.RecordClip"/>) —
    /// it's a continuous cost, so keep it to tester builds.
    /// </summary>
    internal sealed class ClipRecorder : MonoBehaviour
    {
        private byte[][] _ring;
        private int _next;
        private bool _wrapped;
        private readonly object _lock = new object();
        private int _maxW, _quality, _inFlight;
        private float _interval, _lastCapture;
        private bool _flipY;

        /// <summary>The fps actually used (after clamping) — travels with the report so the dashboard plays
        /// the clip at real speed instead of guessing.</summary>
        public int Fps { get; private set; }

        public static ClipRecorder Create()
        {
            var cfg = BugReporter.Config;
            var go = new GameObject("[BugReporter.Clip]");
            DontDestroyOnLoad(go);
            go.hideFlags = HideFlags.HideInHierarchy;
            var r = go.AddComponent<ClipRecorder>();

            int fps = Mathf.Clamp(cfg.ClipFps, 1, 15);
            int frames = Mathf.Clamp(Mathf.Max(1, cfg.ClipSeconds) * fps, fps, 900);
            r._ring = new byte[frames][];
            r._interval = 1f / fps;
            r.Fps = fps;
            r._quality = Mathf.Clamp(cfg.ClipQuality, 1, 100);
            r._maxW = Mathf.Clamp(cfg.ClipMaxWidth, 120, 1280);
            r._flipY = cfg.ClipFlipY;

            if (SystemInfo.supportsAsyncGPUReadback)
            {
                r.StartCoroutine(r.CaptureLoop());
                Debug.Log($"[BugReporter] Clip recording on — last {cfg.ClipSeconds}s at {fps}fps ({frames} frames).");
                if (frames > 300)
                {
                    Debug.LogWarning($"[BugReporter] {frames} frames is a big ring — only the newest " +
                        $"~{Mathf.Max(256 * 1024, cfg.ClipMaxBytes) / 1048576}MB is ever uploaded (the rest just holds RAM, " +
                        $"roughly {frames * 35 / 1024}MB of it). A shorter ClipSeconds gets you the same clip for less.");
                }
            }
            else
            {
                Debug.LogWarning("[BugReporter] Async GPU readback unsupported — clip recording disabled on this device.");
            }
            return r;
        }

        private IEnumerator CaptureLoop()
        {
            var wait = new WaitForEndOfFrame();
            while (true)
            {
                yield return wait;
                if (Time.unscaledTime - _lastCapture < _interval) continue;
                if (_inFlight > 1) continue;          // don't let readbacks pile up under load
                _lastCapture = Time.unscaledTime;
                try { CaptureOne(); }
                catch (Exception e) { Debug.LogWarning($"[BugReporter] Clip capture skipped: {e.Message}"); }
            }
        }

        private void CaptureOne()
        {
            int sw = Screen.width, sh = Screen.height;
            if (sw <= 0 || sh <= 0) return;

            float scale = Mathf.Min(1f, (float)_maxW / sw);
            int tw = Mathf.Max(1, Mathf.RoundToInt(sw * scale));
            int th = Mathf.Max(1, Mathf.RoundToInt(sh * scale));

            var full = RenderTexture.GetTemporary(sw, sh, 0, RenderTextureFormat.ARGB32);
            ScreenCapture.CaptureScreenshotIntoRenderTexture(full);
            var small = RenderTexture.GetTemporary(tw, th, 0, RenderTextureFormat.ARGB32);
            if (_flipY) Graphics.Blit(full, small, new Vector2(1, -1), new Vector2(0, 1));
            else Graphics.Blit(full, small);
            RenderTexture.ReleaseTemporary(full);

            _inFlight++;
            AsyncGPUReadback.Request(small, 0, TextureFormat.RGBA32, req =>
            {
                _inFlight = Mathf.Max(0, _inFlight - 1);
                try
                {
                    if (!req.hasError)
                    {
                        var data = req.GetData<byte>();
                        var jpg = ImageConversion.EncodeNativeArrayToJPG(
                            data, GraphicsFormat.R8G8B8A8_UNorm, (uint)tw, (uint)th, 0, _quality);
                        byte[] bytes = jpg.ToArray();
                        jpg.Dispose();
                        Push(bytes);
                    }
                }
                catch (Exception e) { Debug.LogWarning($"[BugReporter] Clip frame encode failed: {e.Message}"); }
                finally { RenderTexture.ReleaseTemporary(small); }
            });
        }

        private void Push(byte[] frame)
        {
            lock (_lock)
            {
                _ring[_next] = frame;
                _next = (_next + 1) % _ring.Length;
                if (_next == 0) _wrapped = true;
            }
        }

        /// <summary>
        /// Snapshot the buffered frames, oldest-to-newest, into one blob the backend can split:
        /// <c>[uint32 count][uint32 len]×count][frame bytes…]</c> (little-endian). Null when empty.
        /// </summary>
        public byte[] PackLatest()
        {
            var all = new List<byte[]>();
            lock (_lock)
            {
                int count = _wrapped ? _ring.Length : _next;
                int start = _wrapped ? _next : 0;
                for (int i = 0; i < count; i++)
                {
                    byte[] f = _ring[(start + i) % _ring.Length];
                    if (f != null) all.Add(f);
                }
            }
            if (all.Count == 0) return null;

            // Budget from the newest frame backwards. An oversized clip gets the whole report rejected by the
            // server (413) — screenshot and logs with it — so trim here instead, dropping the OLDEST seconds:
            // what happened right before the tester hit Report is the part worth keeping.
            int budget = Mathf.Max(256 * 1024, BugReporter.Config.ClipMaxBytes);
            int total = 4, first = all.Count;
            for (int i = all.Count - 1; i >= 0; i--)
            {
                int cost = all[i].Length + 4;          // frame bytes + its length prefix
                if (total + cost > budget) break;
                total += cost;
                first = i;
            }
            if (first >= all.Count) return null;       // even one frame doesn't fit — send no clip at all

            var frames = all.GetRange(first, all.Count - first);
            if (frames.Count < all.Count)
            {
                Debug.LogWarning($"[BugReporter] Clip trimmed {all.Count}→{frames.Count} frames " +
                    $"(~{frames.Count / Mathf.Max(1, Fps)}s, {total / 1048576f:F1}MB) to fit ClipMaxBytes. " +
                    "Lower ClipSeconds/ClipMaxWidth if you want the full window.");
            }

            using var ms = new MemoryStream(total);
            using var bw = new BinaryWriter(ms);   // BinaryWriter is little-endian on every platform
            bw.Write((uint)frames.Count);
            foreach (var f in frames) bw.Write((uint)f.Length);
            foreach (var f in frames) bw.Write(f);
            bw.Flush();
            return ms.ToArray();
        }

        private void OnDestroy() => StopAllCoroutines();
    }
}

File: flutter-sdk/bug_reporter/lib/src/clip_recorder.dart
import 'dart:async';
import 'dart:typed_data';

import 'package:flutter/foundation.dart';
import 'package:flutter/scheduler.dart';

import 'capture.dart';

/// Rolling buffer of the last few seconds of the app as small JPEG frames. A tick grabs a frame right
/// after the next paint and encodes it in a background isolate; a tick is skipped while the previous frame
/// is still encoding, so a slow phone gets fewer frames instead of a backlog.
class ClipRecorder {
  ClipRecorder({required int seconds, required int fps, required this.maxWidth, required this.quality, required this.maxBytes})
      : fps = fps.clamp(1, 10),
        _ring = List<Uint8List?>.filled((seconds < 1 ? 1 : seconds) * fps.clamp(1, 10), null);

  final int fps;
  final int maxWidth;
  final int quality;
  final int maxBytes;
  final List<Uint8List?> _ring;
  int _next = 0;
  bool _wrapped = false;
  bool _busy = false;
  Timer? _timer;

  void start() {
    _timer ??= Timer.periodic(Duration(milliseconds: 1000 ~/ fps), (_) => _tick());
  }

  void stop() {
    _timer?.cancel();
    _timer = null;
  }

  void _tick() {
    if (_busy) return;
    _busy = true;
    SchedulerBinding.instance.addPostFrameCallback((_) async {
      try {
        final frame = await ScreenCapture.grab(maxWidth: maxWidth);
        if (frame != null) push(await compute(encodeJpeg, (frame, quality.clamp(1, 100))));
      } catch (_) {
        // a skipped frame is fine
      } finally {
        _busy = false;
      }
    });
    SchedulerBinding.instance.ensureVisualUpdate();
  }

  @visibleForTesting
  void push(Uint8List frame) {
    _ring[_next] = frame;
    _next = (_next + 1) % _ring.length;
    if (_next == 0) _wrapped = true;
  }

  /// Oldest-to-newest frames packed as `[u32 count][u32 length × count][frame bytes…]` (little-endian) — the
  /// format the server unpacks. Trimmed from the oldest end to fit [maxBytes]; null when there is nothing.
  Uint8List? packLatest() {
    final all = <Uint8List>[];
    final count = _wrapped ? _ring.length : _next;
    final start = _wrapped ? _next : 0;
    for (var i = 0; i < count; i++) {
      final f = _ring[(start + i) % _ring.length];
      if (f != null) all.add(f);
    }
    if (all.isEmpty) return null;

    final budget = maxBytes < 256 * 1024 ? 256 * 1024 : maxBytes;
    var total = 4, first = all.length;
    for (var i = all.length - 1; i >= 0; i--) {
      final cost = all[i].length + 4;
      if (total + cost > budget) break;
      total += cost;
      first = i;
    }
    if (first >= all.length) return null;
    final frames = all.sublist(first);

    final out = BytesBuilder(copy: false);
    final header = ByteData(4 + 4 * frames.length)..setUint32(0, frames.length, Endian.little);
    for (var i = 0; i < frames.length; i++) {
      header.setUint32(4 + 4 * i, frames[i].length, Endian.little);
    }
    out.add(header.buffer.asUint8List());
    for (final f in frames) {
      out.add(f);
    }
    return out.takeBytes();
  }
}
